### PR TITLE
Feature/20144 Refactor add and edit check window functionality

### PR DESCRIPTION
**Pull request by GeorgeChatzigiannis**

### CODE REVIEW

File: admin/services/check-window.service.js [hunk 1]
```diff
@@ -114,6 +93,88 @@ const checkWindowService = {
...
+    // fetch past current and future check windows
+    const checkWindows = await checkWindowDataService.sqlFindAllCheckWindows(sortField, sortDirection)
+    const pastCheckWindows = []
+    const newCheckWindows = []
+    checkWindows.map(cw => {
+      const adminStartDate = moment(cw.adminStartDate)
+      const checkStartDate = moment(cw.checkStartDate)
+      const checkEndDate = moment(cw.checkEndDate)
+
+      if (moment(checkStartDate).isAfter(moment(checkEndDate))) {
```
Comment (jon-shipley) [inline admin/services/check-window.service.js]: Aren't `checkStartDate` and `checkEndDate` already moment objects?  Why clone them?

File: admin/services/check-window.service.js [hunk 2]
```diff
@@ -114,6 +93,88 @@ const checkWindowService = {
...
+        id: cw.id,
+        name: cw.name,
+        adminStartDate: adminStartDate.format('D MMM YYYY'),
+        checkDates: dateService.formatCheckPeriod(checkStartDate, checkEndDate),
+        isPast
+      }
+
+      if (isPast) {
+        checkWindow.canRemove = false
+        pastCheckWindows.push(checkWindow)
```
Comment (jon-shipley) [inline admin/services/check-window.service.js]: Adding `checkWindow` to different arrays and then concatenting the arrays doesn't actually seem to be required.  The code would be very slightly simpler without that step.  Is there a reason for it?

File: admin/services/check-window.service.js [hunk 3]
```diff
@@ -114,6 +93,88 @@ const checkWindowService = {
...
+      adminStartYear: moment(checkWindow.adminStartDate).format('YYYY'),
+      checkStartDay: moment(checkWindow.checkStartDate).format('D'),
+      checkStartMonth: moment(checkWindow.checkStartDate).format('MM'),
+      checkStartYear: moment(checkWindow.checkStartDate).format('YYYY'),
+      checkEndDay: moment(checkWindow.checkEndDate).format('D'),
+      checkEndMonth: moment(checkWindow.checkEndDate).format('MM'),
+      checkEndYear: moment(checkWindow.checkEndDate).format('YYYY'),
+      existingAdminStartDate: adminStartDate,
+      existingCheckStartDate: checkStartDate,
+      adminIsDisabled: moment(currentDate).isAfter(adminStartDate) ? 1 : 0,
```
Comment (jon-shipley) [inline admin/services/check-window.service.js]: Why not use `true` or `false` and convert them in the UI?

File: admin/services/data-access/check-window.data.service.js [hunk 4]
```diff
@@ -237,6 +237,32 @@ const checkWindowDataService = {
     ]
     const result = await sqlService.query(sql, params)
     return R.head(result)
+  },
+
+  /**
+   * Find all check windows
+   * @param sortBy
+   * @param sortDirection
+   * @return {Object}
```
Comment (jon-shipley) [inline admin/services/data-access/check-window.data.service.js]: This function returns an array of object?  So it should be `@returns [{Object}] - array of checkWindows`

File: admin/views/service-manager/check-windows-form.ejs [hunk 5]
```diff
@@ -33,11 +33,11 @@
                     <span class="error-message"><%= error.checkWindowName %></span>
                     <% } %>
                 </label>
-                <input class="form-control <%= error.checkWindowName ? 'form-control-error' : '' %>"
+                <input class="form-control <%= typeof error !== 'undefined' && error.checkWindowName ? 'form-control-error' : '' %>"
```
Comment (jon-shipley) [inline admin/views/service-manager/check-windows-form.ejs]: With all these checks can't the error object be initialised so it isn't `undefined`?  Would simplify imo.

File: admin/services/check-window.service.js (hunk 1 above)
Comment (GeorgeChatzigiannis) [reply admin/services/check-window.service.js]: @jon-shipley updated

File: admin/services/check-window.service.js (hunk 3 above)
Comment (GeorgeChatzigiannis) [reply admin/services/check-window.service.js]: @jon-shipley updated

File: admin/services/data-access/check-window.data.service.js (hunk 4 above)
Comment (GeorgeChatzigiannis) [reply admin/services/data-access/check-window.data.service.js]: @jon-shipley correct, it's updated

File: admin/views/service-manager/check-windows-form.ejs (hunk 5 above)
Comment (GeorgeChatzigiannis) [reply admin/views/service-manager/check-windows-form.ejs]: @jon-shipley updated

File: admin/services/check-window.service.js [hunk 6]
```diff
@@ -142,6 +142,39 @@ const checkWindowService = {
...
+    }
+    const checkWindow = await checkWindowDataService.sqlFindOneById(id)
+    const currentDate = moment().format('YYYY-MM-D')
+    const adminStartDate = moment(checkWindow.adminStartDate, 'D MM YYYY').format('YYYY-MM-D')
+    const checkStartDate = moment(checkWindow.checkStartDate, 'D MM YYYY').format('YYYY-MM-D')
+
+    return {
+      checkWindowId: id,
+      checkWindowName: checkWindow.name,
+      adminStartDay: moment(checkWindow.adminStartDate).format('D'),
```
Comment (jon-shipley) [inline admin/services/check-window.service.js]: Why are the dates re-parsed every time?  Wouldn't it make sense here to parse them once with moment and _then_  extract the date-parts with format?

File: admin/views/service-manager/check-windows-form.ejs [hunk 7]
```diff
@@ -48,59 +48,59 @@
...
                     </div>
                     <div class="form-group form-group-year">
                         <label class="form-label" for="adminStartYear">Year</label>
-                        <input class="form-control <%= error.adminStartYear ? 'form-control-error' : '' %>"
+                        <input class="form-control <%= error.isError('adminStartYear') ? 'form-control-error' : '' %>"
                                id="adminStartYear"
                                name="adminStartYear"
                                type="number"
-                               value="<%= checkWindowData.adminStartYear %>"
+                               value="<%= typeof checkWindowData !== 'undefined' && checkWindowData.adminStartYear %>"
```
Comment (jon-shipley) [inline admin/views/service-manager/check-windows-form.ejs]: Why do we need all these checks ` typeof checkWindowData !== 'undefined'`?

File: admin/services/check-window.service.js (hunk 6 above)
Comment (GeorgeChatzigiannis) [reply admin/services/check-window.service.js]: @jon-shipley updated

File: admin/views/service-manager/check-windows-form.ejs (hunk 7 above)
Comment (GeorgeChatzigiannis) [reply admin/views/service-manager/check-windows-form.ejs]: @jon-shipley updated

**Pull request merged**